File: Mastering DSA for Developers/Day_Three/HashTable/HashTableFull.js
class Node {
    constructor(key, value) {
        this.key = key;
        this.value = value;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
    }

    insert(key, value) {
        const newNode = new Node(key, value);
        if (!this.head) {
            this.head = newNode;
            return;
        }

        const existing = this.#findNode(key);
        if (existing) {
            existing.value = value;
            return;
        }

        newNode.next = this.head;
        this.head = newNode;
    }

    #findNode(key) {
        if (!this.head) return null;

        let current = this.head;
        while (current) {
            if (current.key === key) return current;
            current = current.next;
        }

        return null;
    }

    find(key) {
        const node = this.#findNode(key);
        return node ? node.value : null;
    }

    *entries() {
        let current = this.head;
        while (current) {
            yield [current.key, current.value];
            current = current.next;
        }
    }
}

class HashTable {
    #keys;
    constructor(size = 3) {
        this.size = size;
        this.table = new Array(size);
        this.#keys = new Set();
        this.count = 0;
    }

    // Better hash function
    #hash(key, size = this.size) {
        let hash = 5381;
        for (let ch of key) {
            hash = (hash * 33) ^ ch.charCodeAt(0);
        }
        return Math.abs(hash) % size;
    }

    #resize(newSize) {
        const oldData = this.table;
        this.size = newSize;
        this.table = new Array(newSize);
        this.count = 0;

        for (let bucket of oldData) {
            if (bucket) {
                for (let [key, value] of bucket.entries()) {
                    this.set(key, value);
                }
            }
        }
    }

    set(key, value) {
        if (this.count / this.size > 0.75) {
            this.#resize(this.size * 2); // Double the size
        }

        const index = this.#hash(key);
        if (!this.table[index]) {
            this.table[index] = new LinkedList();
        }

        const bucket = this.table[index];
        const existing = bucket.find(key);

        if (!existing) {
            this.count++;
            this.#keys.add(key);
        }

        bucket.insert(key, value);
    }

    get(key) {
        const index = this.#hash(key);
        const bucket = this.table[index];

        if (!bucket) return null;
        return bucket.find(key);
    }

    keys() {
        return Array.from(this.#keys);
    }

    values() {
        const values = [];
        for (let bucket of this.table) {
            if (bucket) {
                for (let [_, value] of bucket.entries()) {
                    values.push(value);
                }
            }
        }
        return values;
    }

    entries() {
        const all = [];
        for (let bucket of this.table) {
            if (bucket) {
                for (let entry of bucket.entries()) {
                    all.push(entry);
                }
            }
        }
        return all;
    }
}

const hashTable = new HashTable();
hashTable.set('name', 'John');
hashTable.set('age', 20);
hashTable.set('city', 'New York');
hashTable.set('country', 'USA');
hashTable.set('phone', '1234567890');
hashTable.set('name', 'Jane');

console.log(hashTable.keys());
console.log(hashTable.values());
console.log(hashTable.entries());

File: Mastering DSA for Developers/day_3/HashTable/HashTableFull.py
class Node:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def insert(self, key, value):
        existing_node = self._find_node(key)
        if existing_node:
            existing_node.value = value
            return

        new_node = Node(key, value)
        new_node.next = self.head
        self.head = new_node

    def _find_node(self, key):
        current = self.head
        while current:
            if current.key == key:
                return current
            current = current.next
        return None

    def find(self, key):
        node = self._find_node(key)
        return node.value if node else None

    def entries(self):
        current = self.head
        while current:
            yield (current.key, current.value)
            current = current.next


class HashTable:
    def __init__(self, size=3):
        self.size = size
        self.table = [None] * size
        self._keys = set()
        self.count = 0

    def _hash(self, key, size=None):
        if size is None:
            size = self.size

        has_value = 5381
        for ch in key:
            has_value = (has_value * 33) ^ ord(ch)
        return abs(has_value) % size

    def _resize(self, new_size):
        pass

    def set(self, key, value):

        index = self._hash(key)
        if self.table[index] is None:
            self.table[index] = LinkedList()

        bucket = self.table[index]
        if bucket.find(key) is None:
            self.count += 1
            self._keys.add(key)

        bucket.insert(key, value)

    def get(self, key):
        index = self._hash(key)
        bucket = self.table[index]
        if not bucket:
            return None
        return bucket.find(key)

    def keys(self):
        return list(self._keys)

    def values(self):
        values = []
        for bucket in self.table:
            if bucket:
                for _, value in bucket.entries():
                    values.append(value)

        return values

    def entries(self):
        all_entries = []
        for bucket in self.table:
            if bucket:
                for entry in bucket.entries():
                    all_entries.append(entry)

        return all_entries


hash_table = HashTable()

hash_table.set("name", "Md Naimur Rahman")
hash_table.set("phone", "01521438923")
hash_table.set("city", "Dhaka")
print(hash_table.entries())
print(hash_table.keys())
print(hash_table.values())
print(hash_table.get("name"))
